tket/src/Characterisation/Cycles.cpp

```diff
@@ -79,49 +79,119 @@ void CycleFinder::erase_keys(
   for (const unsigned& key : all_erased) erase_from.erase(key);
 }
 
-// Adds a new cycle to bt.key_to_cycle
+// Adds a new cycle to this->cycle_history.key_to_cycle
 std::pair<unsigned, std::set<unsigned>> CycleFinder::make_cycle(
     const Vertex& v, const EdgeVec& out_edges, const CutFrontier& cut) {
   // Make a new boundary, add it to "all_boundaries", update "boundary_key" map
   std::vector<edge_pair_t> new_cycle_boundary;
-  std::vector<UnitID> new_boundary_uids;
+  std::set<UnitID> new_boundary_uids;
   std::set<unsigned> old_boundary_keys;
-  std::set<unsigned> banned_keys;
+  std::set<unsigned> not_mergeable_keys;
+  std::set<UnitID> not_mergeable_uids;
 
   // Get UnitID, add to uids for new boundary, keep note of all old boundary
   // keys for merging set new boundary key, add new edges to new boundary update
   // cycle out edges if the edge can't be found
+
   for (const Edge& e : out_edges) {
     UnitID uid = unitid_from_unit_frontier(cut.u_frontier, e);
-    new_boundary_uids.push_back(uid);
-    old_boundary_keys.insert(bt.uid_to_key[uid]);
-
+    new_boundary_uids.insert(uid);
+    old_boundary_keys.insert(this->cycle_history.uid_to_key[uid]);
     // boundary key associated with given edge e not included
     // i.e. e's associated in edge isn't a cycle out edge
     // i.e. some other non-cycle gate has been passed
+
+    // if the edge going into the vertex is not in a previous cycle
+    // then it can't be merged with that cycle
     if (cycle_out_edges.find(circ.get_last_edge(v, e)) ==
         cycle_out_edges.end()) {
-      banned_keys.insert(bt.uid_to_key[uid]);
+      not_mergeable_keys.insert(this->cycle_history.uid_to_key[uid]);
+      // not_mergeable_uids.insert(uid);
     }
-    bt.uid_to_key[uid] = bt.key;
+    this->cycle_history.uid_to_key[uid] = this->cycle_history.key;
     new_cycle_boundary.push_back({circ.get_last_edge(v, e), e});
     update_cycle_out_edges(uid, e);
   }
 
+  for (const unsigned& key : old_boundary_keys) {
+    for (const UnitID& uid : this->cycle_history.history[key]) {
+      if (not_mergeable_uids.find(uid) != not_mergeable_uids.end()) {
+        not_mergeable_keys.insert(key);
+        // break;
+        continue;
```

Comment:
This `continue;` is unnecessary isn't it?

Reply:
Removed